### PR TITLE
Add support for GraphQL 3 and 4

### PR DESCRIPTION
**Pull request by kevingosse**
Contributes to https://github.com/DataDog/dd-trace-dotnet/issues/1721

### CODE REVIEW

File: integrations.json [hunk 1]
```diff
@@ -1237,7 +1237,30 @@
           "minimum_major": 2,
           "minimum_minor": 3,
           "minimum_patch": 0,
-          "maximum_major": 2,
+          "maximum_major": 3,
+          "maximum_minor": 65535,
+          "maximum_patch": 65535
+        },
+        "wrapper": {
+          "assembly": "Datadog.Trace, Version=1.28.5.0, Culture=neutral, PublicKeyToken=def86d061d0d2eeb",
```
Comment (andrewlock) [inline integrations.json:1245]: Needs updating

File: test/test-applications/integrations/Samples.GraphQL3/Properties/launchSettings.json [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+        "DD_DOTNET_TRACER_HOME": "$(ProjectDir)$(OutputPath)profiler-lib",
+        "DD_INTEGRATIONS": "$(ProjectDir)$(OutputPath)profiler-lib\\integrations.json",
+        "DD_VERSION": "1.0.0"
+      },
+      "applicationUrl": "http://localhost:54568/",
+      "nativeDebugging": true
+    }
+  }
+}
```
Comment (andrewlock) [inline test/test-applications/integrations/Samples.GraphQL3/Properties/launchSettings.json:54]: I don't think we need the IIS profile
```suggestion
{
  "profiles": {
    "Samples.GraphQL3": {
      "commandName": "Project",
      "launchBrowser": false,
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development",

        "COR_ENABLE_PROFILING": "1",
        "COR_PROFILER": "{846F5F1C-F9AE-4B07-969E-05C26BC060D8}",
        "COR_PROFILER_PATH": "$(ProjectDir)$(OutputPath)profiler-lib\\Datadog.Trace.ClrProfiler.Native.dll",

        "CORECLR_ENABLE_PROFILING": "1",
        "CORECLR_PROFILER": "{846F5F1C-F9AE-4B07-969E-05C26BC060D8}",
        "CORECLR_PROFILER_PATH": "$(ProjectDir)$(OutputPath)profiler-lib\\Datadog.Trace.ClrProfiler.Native.dll",

        "DD_DOTNET_TRACER_HOME": "$(ProjectDir)$(OutputPath)profiler-lib",
        "DD_INTEGRATIONS": "$(ProjectDir)$(OutputPath)profiler-lib\\integrations.json",
        "DD_VERSION": "1.0.0"
      },
      "applicationUrl": "http://localhost:54568/",
      "nativeDebugging": true
    }
  }
}
```

File: test/test-applications/integrations/Samples.GraphQL4/Properties/launchSettings.json [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+        "DD_DOTNET_TRACER_HOME": "$(ProjectDir)$(OutputPath)profiler-lib",
+        "DD_INTEGRATIONS": "$(ProjectDir)$(OutputPath)profiler-lib\\integrations.json",
+        "DD_VERSION": "1.0.0"
+      },
+      "applicationUrl": "http://localhost:54568/",
+      "nativeDebugging": true
+    }
+  }
+}
```
Comment (andrewlock) [inline test/test-applications/integrations/Samples.GraphQL4/Properties/launchSettings.json:54]: Go away IIS
```suggestion
{
  "profiles": {
    "Samples.GraphQL4": {
      "commandName": "Project",
      "launchBrowser": false,
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development",

        "COR_ENABLE_PROFILING": "1",
        "COR_PROFILER": "{846F5F1C-F9AE-4B07-969E-05C26BC060D8}",
        "COR_PROFILER_PATH": "$(ProjectDir)$(OutputPath)profiler-lib\\Datadog.Trace.ClrProfiler.Native.dll",

        "CORECLR_ENABLE_PROFILING": "1",
        "CORECLR_PROFILER": "{846F5F1C-F9AE-4B07-969E-05C26BC060D8}",
        "CORECLR_PROFILER_PATH": "$(ProjectDir)$(OutputPath)profiler-lib\\Datadog.Trace.ClrProfiler.Native.dll",

        "DD_DOTNET_TRACER_HOME": "$(ProjectDir)$(OutputPath)profiler-lib",
        "DD_INTEGRATIONS": "$(ProjectDir)$(OutputPath)profiler-lib\\integrations.json",
        "DD_VERSION": "1.0.0"
      },
      "applicationUrl": "http://localhost:54568/",
      "nativeDebugging": true
    }
  }
}
```

File: test/test-applications/integrations/Samples.GraphQL4/StarWars/HumanInputType.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    public class HumanInputType : InputObjectGraphType
+    {
+        public HumanInputType()
+        {
+            Name = "HumanInput";
+            Field<NonNullGraphType<StringGraphType>>("name");
+            Field<StringGraphType>("homePlanet");
+        }
+    }
+}
```
Comment (andrewlock) [inline test/test-applications/integrations/Samples.GraphQL4/StarWars/HumanInputType.cs:14]: And if these are shared between 3 + 4, could link them too

**Review by andrewlock — APPROVED** [review]
LGTM, just a few nit comments. Nice work 👍

**Review by zacharycmontoya — APPROVED** [review]
LGTM

File: src/Datadog.Trace/ClrProfiler/AutoInstrumentation/GraphQL/ValidateAsync4Integration.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            if (state.Scope is null)
+            {
+                return validationResult;
+            }
+
+            try
+            {
+                if (exception != null)
+                {
+                    scope.Span?.SetException(exception);
```
[the remark this answers is not in the document]
Comment (kevingosse) [reply src/Datadog.Trace/ClrProfiler/AutoInstrumentation/GraphQL/ValidateAsync4Integration.cs:78]: I guess. Still too scared to remove it 😛

File: test/test-applications/integrations/Samples.GraphQL3/GraphQLUserContext.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+using System.Security.Claims;
+
+namespace Samples.GraphQL3
+{
+    public class GraphQLUserContext
+    {
+        public ClaimsPrincipal User { get; set; }
+    }
+}
```
[the remark this answers is not in the document]
Comment (kevingosse) [reply test/test-applications/integrations/Samples.GraphQL3/GraphQLUserContext.cs:9]: There are subtle differences (for instance `StarWarsData` uses `Apply` in GraphQL3 but that method has been removed in GraphQL4. It's probably possible to rewrite them with a common set of APIs but I don't think it's worth the trouble

File: test/test-applications/integrations/Samples.GraphQL4/Properties/launchSettings.json (hunk 3 above)
Comment (kevingosse) [reply test/test-applications/integrations/Samples.GraphQL4/Properties/launchSettings.json]: Done in https://github.com/DataDog/dd-trace-dotnet/pull/1751/commits/7643a3a2584b683f3567f12ddeeb837d7931a8c5

File: test/test-applications/integrations/Samples.GraphQL3/Properties/launchSettings.json (hunk 2 above)
Comment (kevingosse) [reply test/test-applications/integrations/Samples.GraphQL3/Properties/launchSettings.json]: Done in https://github.com/DataDog/dd-trace-dotnet/pull/1751/commits/7643a3a2584b683f3567f12ddeeb837d7931a8c5

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/GraphQLTests.cs [hunk 7]
```diff
@@ -73,6 +73,11 @@ public GraphQLTests(ITestOutputHelper output)
         [InlineData(true)]
         public void SubmitsTraces(bool enableCallTarget)
         {
+            if (!enableCallTarget && _disableCallsite)
```
[the remark this answers is not in the document]
Comment (kevingosse) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/GraphQLTests.cs]: Done in https://github.com/DataDog/dd-trace-dotnet/pull/1751/commits/7643a3a2584b683f3567f12ddeeb837d7931a8c5

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/GraphQLTests.cs [hunk 8]
```diff
@@ -18,52 +21,49 @@
...
-            // FAILURE: query fails 'execute' step
-            CreateGraphQLRequestsAndExpectations(url: "/graphql", httpMethod: "POST", resourceName: "Subscription NotImplementedSub", graphQLRequestBody: @"{""query"":""subscription NotImplementedSub{throwNotImplementedException{name}}""}", graphQLOperationType: "Subscription", graphQLOperationName: "NotImplementedSub", graphQLSource: "subscription NotImplementedSub{throwNotImplementedException{name}}", failsExecution: true);
+        private List<RequestInfo> _requests;
+        private List<WebServerSpanExpectation> _expectations;
+        private int _expectedGraphQLValidateSpanCount;
+        private int _expectedGraphQLExecuteSpanCount;
 
-            // TODO: When parse is implemented, add a test that fails 'resolve' step
+        public GraphQLTests(ITestOutputHelper output)
+            : this("GraphQL", output)
```
[the remark this answers is not in the document]
Comment (kevingosse) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/GraphQLTests.cs]: Done in https://github.com/DataDog/dd-trace-dotnet/pull/1751/commits/7643a3a2584b683f3567f12ddeeb837d7931a8c5

File: integrations.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (kevingosse) [reply integrations.json]: Done

**Pull request merged**